File: clients/go/consensus/encode.go
package consensus

import "encoding/binary"

// Target (raw bytes), and Nonce (8-byte little-endian).
func BlockHeaderBytes(header BlockHeader) []byte {
	out := make([]byte, 0, 4+32+32+8+32+8)
	var tmp4 [4]byte
	var tmp8 [8]byte

	binary.LittleEndian.PutUint32(tmp4[:], header.Version)
	out = append(out, tmp4[:]...)
	out = append(out, header.PrevBlockHash[:]...)
	out = append(out, header.MerkleRoot[:]...)
	binary.LittleEndian.PutUint64(tmp8[:], header.Timestamp)
	out = append(out, tmp8[:]...)
	out = append(out, header.Target[:]...)
	binary.LittleEndian.PutUint64(tmp8[:], header.Nonce)
	out = append(out, tmp8[:]...)
	return out
}

// TxOutputBytes serializes a TxOutput into its canonical byte representation.
//
// It encodes Value as 8-byte little-endian, CovenantType as 2-byte little-endian,
// then the length of CovenantData using CompactSize followed by the CovenantData bytes.
// The resulting slice is the concatenation of those fields in that order.
func TxOutputBytes(o TxOutput) []byte {
	out := make([]byte, 0, 8+2+9+len(o.CovenantData))
	var tmp8 [8]byte
	binary.LittleEndian.PutUint64(tmp8[:], o.Value)
	out = append(out, tmp8[:]...)
	var tmp2 [2]byte
	binary.LittleEndian.PutUint16(tmp2[:], o.CovenantType)
	out = append(out, tmp2[:]...)
	out = append(out, CompactSize(len(o.CovenantData)).Encode()...)
	out = append(out, o.CovenantData...)
	return out
}

// WitnessItemBytes serializes a WitnessItem into its wire format.
// The encoding is: SuiteID (1 byte), Pubkey length encoded as CompactSize, Pubkey bytes,
// Signature length encoded as CompactSize, then Signature bytes.
// The returned slice contains the complete serialized witness item.
func WitnessItemBytes(w WitnessItem) []byte {
	out := make([]byte, 0, 1+9+len(w.Pubkey)+9+len(w.Signature))
	out = append(out, w.SuiteID)
	out = append(out, CompactSize(len(w.Pubkey)).Encode()...)
	out = append(out, w.Pubkey...)
	out = append(out, CompactSize(len(w.Signature)).Encode()...)
	out = append(out, w.Signature...)
	return out
}

// WitnessBytes serializes a witness section into a byte slice.
// It encodes the number of witness items using CompactSize and then appends each witness item serialized by WitnessItemBytes, returning the concatenated bytes.
func WitnessBytes(w WitnessSection) []byte {
	out := make([]byte, 0, 9)
	out = append(out, CompactSize(len(w.Witnesses)).Encode()...)
	for _, item := range w.Witnesses {
		out = append(out, WitnessItemBytes(item)...)
	}
	return out
}

// TxNoWitnessBytes serializes a transaction excluding its witness section into a byte slice.
//
// The serialized layout is:
// - Version (4 bytes, little-endian)
// - TxNonce (8 bytes, little-endian)
// - Inputs count (CompactSize) followed by each input:
//   - PrevTxid (32 bytes)
//   - PrevVout (4 bytes, little-endian)
//   - ScriptSig length (CompactSize) and ScriptSig bytes
//   - Sequence (4 bytes, little-endian)
//
// - Outputs count (CompactSize) followed by each output serialized by TxOutputBytes
// - Locktime (4 bytes, little-endian)
func TxNoWitnessBytes(tx *Tx) []byte {
	out := make([]byte, 0, 4+8)
	var tmp4 [4]byte
	binary.LittleEndian.PutUint32(tmp4[:], tx.Version)
	out = append(out, tmp4[:]...)
	var tmp8 [8]byte
	binary.LittleEndian.PutUint64(tmp8[:], tx.TxNonce)
	out = append(out, tmp8[:]...)

	out = append(out, CompactSize(len(tx.Inputs)).Encode()...)
	for _, in := range tx.Inputs {
		out = append(out, in.PrevTxid[:]...)
		binary.LittleEndian.PutUint32(tmp4[:], in.PrevVout)
		out = append(out, tmp4[:]...)
		out = append(out, CompactSize(len(in.ScriptSig)).Encode()...)
		out = append(out, in.ScriptSig...)
		binary.LittleEndian.PutUint32(tmp4[:], in.Sequence)
		out = append(out, tmp4[:]...)
	}

	out = append(out, CompactSize(len(tx.Outputs)).Encode()...)
	for _, o := range tx.Outputs {
		out = append(out, TxOutputBytes(o)...)
	}

	binary.LittleEndian.PutUint32(tmp4[:], tx.Locktime)
	out = append(out, tmp4[:]...)
	return out
}

// TxBytes serializes tx into its complete binary representation including its witness section.
// The returned slice contains the transaction fields followed by the serialized witness data.
func TxBytes(tx *Tx) []byte {
	out := TxNoWitnessBytes(tx)
	out = append(out, WitnessBytes(tx.Witness)...)
	return out
}

// BlockBytes serializes a Block into its canonical byte representation.
// The result is the concatenation of the serialized block header, the number of transactions encoded as a CompactSize, and each transaction serialized (including witnesses).
func BlockBytes(block *Block) []byte {
	out := make([]byte, 0, 64)
	out = append(out, BlockHeaderBytes(block.Header)...)
	out = append(out, CompactSize(len(block.Transactions)).Encode()...)
	for _, tx := range block.Transactions {
		out = append(out, TxBytes(&tx)...)
	}
	return out
}

File: clients/rust/crates/rubin-consensus/src/encode.rs
use crate::{BlockHeader, Tx, TxOutput, WitnessItem, WitnessSection, compact_size_encode};

/// Serialize a BlockHeader into its 116-byte canonical header representation.
///
/// The returned array contains, in order:
/// - version as little-endian (bytes 0..4)
/// - previous block hash (bytes 4..36)
/// - merkle root (bytes 36..68)
/// - timestamp as little-endian (bytes 68..76)
/// - target (bytes 76..108)
/// - nonce as little-endian (bytes 108..116)
///
/// # Examples
///
/// ```
/// let header = BlockHeader {
///     version: 1,
///     prev_block_hash: [0u8; 32],
///     merkle_root: [0u8; 32],
///     timestamp: 0u64,
///     target: [0u8; 32],
///     nonce: 0u64,
/// };
/// let bytes = block_header_bytes(&header);
/// assert_eq!(bytes.len(), 116);
/// assert_eq!(&bytes[4..36], &header.prev_block_hash);
/// ```
pub fn block_header_bytes(h: &BlockHeader) -> [u8; 116] {
    let mut out = [0u8; 116];
    out[0..4].copy_from_slice(&h.version.to_le_bytes());
    out[4..36].copy_from_slice(&h.prev_block_hash);
    out[36..68].copy_from_slice(&h.merkle_root);
    out[68..76].copy_from_slice(&h.timestamp.to_le_bytes());
    out[76..108].copy_from_slice(&h.target);
    out[108..116].copy_from_slice(&h.nonce.to_le_bytes());
    out
}

/// Serialize a transaction output including its covenant data into a byte vector.
///
/// The resulting byte vector contains, in order:
/// 1. `value` as little-endian bytes,
/// 2. `covenant_type` as little-endian bytes,
/// 3. the compact-size-encoded length of `covenant_data`,
/// 4. the raw `covenant_data` bytes.
///
/// # Examples
///
/// ```
/// use crate::{TxOutput, tx_output_bytes};
///
/// let output = TxOutput {
///     value: 1u64,
///     covenant_type: 2u32,
///     covenant_data: vec![0xAA, 0xBB],
/// };
///
/// let bytes = tx_output_bytes(&output);
/// assert_eq!(&bytes[0..8], &1u64.to_le_bytes());
/// ```
pub fn tx_output_bytes(output: &TxOutput) -> Vec<u8> {
    let mut out = Vec::new();
    out.extend_from_slice(&output.value.to_le_bytes());
    out.extend_from_slice(&output.covenant_type.to_le_bytes());
    out.extend_from_slice(&compact_size_encode(output.covenant_data.len() as u64));
    out.extend_from_slice(&output.covenant_data);
    out
}

/// Serializes a witness item into a length-prefixed byte sequence.
///
/// The output contains, in order: the single-byte `suite_id`, the compact-size-encoded
/// length of `pubkey` followed by `pubkey` bytes, then the compact-size-encoded length
/// of `signature` followed by `signature` bytes.
///
/// # Parameters
///
/// - `item`: the witness item to serialize.
///
/// # Returns
///
/// A `Vec<u8>` containing the serialized witness item.
///
/// # Examples
///
/// ```
/// let item = WitnessItem {
///     suite_id: 1,
///     pubkey: vec![0x02, 0x03, 0x04],
///     signature: vec![0xaa, 0xbb],
/// };
/// let bytes = witness_item_bytes(&item);
/// // suite_id
/// assert_eq!(bytes[0], 1);
/// // pubkey length (compact size) then pubkey bytes
/// assert!(bytes.windows(3).any(|w| w == [0x03, 0x02, 0x03] ) || bytes.contains(&0x02));
/// ```
pub fn witness_item_bytes(item: &WitnessItem) -> Vec<u8> {
    let mut out = Vec::new();
    out.push(item.suite_id);
    out.extend_from_slice(&compact_size_encode(item.pubkey.len() as u64));
    out.extend_from_slice(&item.pubkey);
    out.extend_from_slice(&compact_size_encode(item.signature.len() as u64));
    out.extend_from_slice(&item.signature);
    out
}

/// Serializes a witness section into bytes.
///
/// The output is the compact-size encoding of the number of witness items followed by each
/// witness item serialized in order.
///
/// # Examples
///
/// ```
/// let w = WitnessSection { witnesses: vec![] };
/// let bytes = witness_bytes(&w);
/// assert_eq!(bytes, compact_size_encode(0));
/// ```
pub fn witness_bytes(w: &WitnessSection) -> Vec<u8> {
    let mut out = Vec::new();
    out.extend_from_slice(&compact_size_encode(w.witnesses.len() as u64));
    for item in &w.witnesses {
        out.extend_from_slice(&witness_item_bytes(item));
    }
    out
}

/// Serialize a transaction excluding its witness section.
///
/// Produces the canonical byte serialization consisting of: version, tx_nonce,
/// compact-size-prefixed inputs (each as prev_txid, prev_vout, script_sig length + script_sig, sequence),
/// compact-size-prefixed outputs (each serialized via `tx_output_bytes`), and locktime.
///
/// # Returns
///
/// A byte vector containing the transaction serialization without witness data.
///
/// # Examples
///
/// ```no_run
/// // Build or obtain a `Tx` value named `tx`, then:
/// let bytes = tx_no_witness_bytes(&tx);
/// assert!(bytes.len() >= 8); // at least version and tx_nonce
/// ```
pub fn tx_no_witness_bytes(tx: &Tx) -> Vec<u8> {
    let mut out = Vec::new();
    out.extend_from_slice(&tx.version.to_le_bytes());
    out.extend_from_slice(&tx.tx_nonce.to_le_bytes());
    out.extend_from_slice(&compact_size_encode(tx.inputs.len() as u64));
    for input in &tx.inputs {
        out.extend_from_slice(&input.prev_txid);
        out.extend_from_slice(&input.prev_vout.to_le_bytes());
        out.extend_from_slice(&compact_size_encode(input.script_sig.len() as u64));
        out.extend_from_slice(&input.script_sig);
        out.extend_from_slice(&input.sequence.to_le_bytes());
    }
    out.extend_from_slice(&compact_size_encode(tx.outputs.len() as u64));
    for output in &tx.outputs {
        out.extend_from_slice(&tx_output_bytes(output));
    }
    out.extend_from_slice(&tx.locktime.to_le_bytes());
    out
}

/// Serialize a transaction into a byte vector including its witness section.
///
/// # Returns
///
/// A `Vec<u8>` containing the complete serialized transaction (no-witness serialization followed by witness data).
///
/// # Examples
///
/// ```
/// let bytes = tx_bytes(&tx); // `tx` is a `Tx`
/// assert!(bytes.len() > 0);
/// ```
pub fn tx_bytes(tx: &Tx) -> Vec<u8> {
    let mut out = tx_no_witness_bytes(tx);
    out.extend_from_slice(&witness_bytes(&tx.witness));
    out
}
